### PR TITLE
fix(DataStore): Generate mutation events for associatedModels

### PR DESCRIPTION
**Pull request by wooj2**
This PR addresses issue [#933](https://github.com/aws-amplify/amplify-ios/issues/933)

This PR relies on:
~~[965](https://github.com/aws-amplify/amplify-ios/pull/965) - Addressing columns~~ -- pushed to mainline
[986](https://github.com/aws-amplify/amplify-ios/pull/986) - minor refactor

(For reference only, the precursor to this pr is located here: https://github.com/aws-amplify/amplify-ios/pull/979 )

Did a high level overview/review with @palpatim -- two items that came up:

**Stress test case**
- See unit test `testStressDeleteTopLevelMultiLevelHasManyRelationship`

**Edge case with regards to foreign key constraints**
Tested an edge case where we have a `Blog` -> `Post` - > `Comment`
Steps:
1.  Updated the storageEngine's code to not send sync deletions for Post
```
+++ b/AmplifyPlugins/DataStore/AWSDataStoreCategoryPlugin/Storage/StorageEngine.swift
@@ -432,6 +432,12 @@ final class StorageEngine: StorageEngineBehavior {
         }
         var mutationEvents: Set<Model.Identifier> = []
         for associatedModel in associatedModels {
+            if associatedModel.modelName == "Post" {
+                print("SKIPPING Post!!")
+                continue
+            }
+
```
2. Created 1 blog, 2 posts, and 4 comments
3. Delete the Blog ( notice that only 5 delete mutation events were sent out )
4.  `clear()` and `start()` -- effectively forcing the initial sync operation to run.

**Observed**: ReconcileAndLocalSaveOperation fails with a Foreign Key constraint coming from SQLite3.  This is expected since its parent Blog(owner), no longer exists.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: AmplifyPlugins/DataStore/AWSDataStoreCategoryPlugin/Storage/StorageEngine.swift [hunk 1]
```diff
@@ -415,6 +422,43 @@ final class StorageEngine: StorageEngineBehavior {
...
 
+    @available(iOS 13.0, *)
+    private func syncDeletions(of associatedModels: [Model],
+                               syncEngine: RemoteSyncEngineBehavior,
+                               completion: @escaping DataStoreCallback<Void>) {
+        guard !associatedModels.isEmpty else {
+            completion(.successfulVoid)
+            return
+        }
+        var mutationEvents: Set<Model.Identifier> = []
```
Comment (palpatim) [inline AmplifyPlugins/DataStore/AWSDataStoreCategoryPlugin/Storage/StorageEngine.swift:433]: 1. Any reason we're not just using a counter of successful mutation events?
2. Is there a data race here, or is the callback to `submitToSyncEngine` guaranteed to always execute on the same serial DispatchQueue, so that this Set is not modified except on that queue?

File: AmplifyPlugins/DataStore/AWSDataStoreCategoryPlugin/Storage/StorageEngine.swift [hunk 2]
```diff
@@ -415,6 +422,43 @@ final class StorageEngine: StorageEngineBehavior {
...
+            do {
+                mutationEvent = try MutationEvent(untypedModel: associatedModel, mutationType: .delete)
+            } catch {
+                let dataStoreError = DataStoreError(error: error)
+                completion(.failure(dataStoreError))
+                return
+            }
+
+            let mutationEventCallback: DataStoreCallback<MutationEvent> = { result in
+                switch result {
```
Comment (palpatim) [inline AmplifyPlugins/DataStore/AWSDataStoreCategoryPlugin/Storage/StorageEngine.swift:445]: I note that we're invoking a completion on failure, but continuing to process events. Continuing makes sense, but it's a little weird to be invoking a completion before things have actually, you know, completed. :) Should we remember the error and then resolve the completion after all the associated models have been processed?

File: AmplifyPlugins/DataStore/AWSDataStoreCategoryPlugin/Storage/StorageEngine.swift [hunk 3]
```diff
@@ -202,10 +202,12 @@ final class StorageEngine: StorageEngineBehavior {
                           modelSchema: ModelSchema,
                           withId id: Model.Identifier,
                           completion: @escaping (DataStoreResult<M?>) -> Void) {
-        let transactionResult = queryAndDeleteTransaction(modelType, modelSchema: modelSchema, predicate: field("id").eq(id))
+        let transactResult = queryAndDeleteTransaction(modelType, modelSchema: modelSchema, predicate: field("id").eq(id))
+        let mTransactionResult = collapseMResult(transactResult)
```
[the remark this answers is not in the document]
Comment (wooj2) [reply AmplifyPlugins/DataStore/AWSDataStoreCategoryPlugin/Storage/StorageEngine.swift:206]: tldr;  - I'll switch it back to `transactionResult`

Yeah, good point, I can change this back to `transactionResult`.  I typically change the name of a variable if i intend to use it differently in efforts of forcing myself to look at every single place that this variable is used, to ensure it's being used correctly.  Previously, we were doing a `switch` on `transactionResult`, so I changed it to `transactResult` to make sure that `transactionResult` wasn't being used anywhere :).

File: AmplifyPlugins/DataStore/AWSDataStoreCategoryPlugin/Storage/StorageEngine.swift [hunk 4]
```diff
@@ -202,10 +202,12 @@ final class StorageEngine: StorageEngineBehavior {
                           modelSchema: ModelSchema,
                           withId id: Model.Identifier,
                           completion: @escaping (DataStoreResult<M?>) -> Void) {
-        let transactionResult = queryAndDeleteTransaction(modelType, modelSchema: modelSchema, predicate: field("id").eq(id))
+        let transactResult = queryAndDeleteTransaction(modelType, modelSchema: modelSchema, predicate: field("id").eq(id))
```
[the remark this answers is not in the document]
Comment (wooj2) [reply AmplifyPlugins/DataStore/AWSDataStoreCategoryPlugin/Storage/StorageEngine.swift:205]: Will handle as part of comment above.

File: AmplifyPlugins/DataStore/AWSDataStoreCategoryPlugin/Storage/StorageEngine+DeleteTransaction.swift [hunk 5]
```diff
@@ -43,18 +47,72 @@ extension StorageEngine {
             return .failure(causedBy: error)
         }
 
-        return collapseResults(queryResult: queriedResult, deleteResult: deletedResult)
+        return collapseResults(queryResult: queriedResult, deleteResult: deletedResult, associatedModels: associatedModels)
```
[the remark this answers is not in the document]
Comment (wooj2) [reply AmplifyPlugins/DataStore/AWSDataStoreCategoryPlugin/Storage/StorageEngine+DeleteTransaction.swift:50]: I believe you are right, it is not synchronous, but i believe this is working because the transaction blocks until the wrapped operations are done.

File: AmplifyPlugins/DataStore/AWSDataStoreCategoryPlugin/Storage/StorageEngine+DeleteTransaction.swift [hunk 6]
```diff
@@ -43,18 +47,72 @@ extension StorageEngine {
...
-    func collapseResults<M: Model>(queryResult: DataStoreResult<[M]>?,
-                                   deleteResult: DataStoreResult<[M]>?) -> DataStoreResult<[M]> {
+    func queryAssociatedModels(modelName: ModelName, associatedField: ModelField, ids: [Model.Identifier]) -> [Model] {
+        guard let modelType = ModelRegistry.modelType(from: modelName) else {
+            log.error("Failed to lookup \(modelName)")
+            return []
+        }
+
+        var queriedModels: [Model] = []
+        let maxNumberOfPredicates = 950
```
[the remark this answers is not in the document]
Comment (wooj2) [reply AmplifyPlugins/DataStore/AWSDataStoreCategoryPlugin/Storage/StorageEngine+DeleteTransaction.swift:80]: Any recommendations on how to change this?

File: AmplifyPlugins/DataStore/AWSDataStoreCategoryPlugin/Storage/StorageEngine+DeleteTransaction.swift [hunk 7]
```diff
@@ -13,13 +13,17 @@ import AWSPluginsCore
...
-                                             predicate: QueryPredicate) -> DataStoreResult<[M]> {
+                                             predicate: QueryPredicate) -> DataStoreResult<([M], [Model])> {
         var queriedResult: DataStoreResult<[M]>?
         var deletedResult: DataStoreResult<[M]>?
+        var associatedModels: [Model] = []
 
         let queryCompletionBlock: DataStoreCallback<[M]> = { queryResult in
             queriedResult = queryResult
-            if case .success = queryResult {
+            if case .success(let queriedModels) = queryResult {
```
[the remark this answers is not in the document]
Comment (wooj2) [reply AmplifyPlugins/DataStore/AWSDataStoreCategoryPlugin/Storage/StorageEngine+DeleteTransaction.swift:23]: updated.

File: AmplifyPlugins/DataStore/AWSDataStoreCategoryPlugin/Storage/StorageEngine.swift (hunk 3 above)
[the remark this answers is not in the document]
Comment (wooj2) [reply AmplifyPlugins/DataStore/AWSDataStoreCategoryPlugin/Storage/StorageEngine.swift:206]: Renamed to:
```
transactionResult
modelsFromTransactionResult
associatedModelsFromTransactionResult
```

File: AmplifyPlugins/DataStore/AWSDataStoreCategoryPlugin/Storage/StorageEngine+DeleteTransaction.swift [hunk 8]
```diff
@@ -43,18 +47,72 @@ extension StorageEngine {
...
+            let sempahore = DispatchSemaphore(value: 0)
+            storageAdapter.query(untypedModel: modelType, predicate: groupedQueryPredicates) { result in
+                switch result {
+                case .success(let models):
+                    queriedModels.append(contentsOf: models)
+
+                case .failure(let error):
+                    log.error("Failed to query \(modelType) on mutation event generation: \(error)")
+                }
+                sempahore.signal()
```
[the remark this answers is not in the document]
Comment (wooj2) [reply AmplifyPlugins/DataStore/AWSDataStoreCategoryPlugin/Storage/StorageEngine+DeleteTransaction.swift]: Added a defer block w/ signaling the semaphore to the beginning of the for loop.. i think this is what you were referring to?

File: AmplifyPlugins/DataStore/AWSDataStoreCategoryPlugin/Storage/StorageEngine+DeleteTransaction.swift (hunk 5 above)
[the remark this answers is not in the document]
Comment (palpatim) [reply AmplifyPlugins/DataStore/AWSDataStoreCategoryPlugin/Storage/StorageEngine+DeleteTransaction.swift:51]: Yep, [`transaction` is synchronous](https://github.com/stephencelis/SQLite.swift/blob/master/Documentation/Index.md#transactions-and-savepoints) Worry resolved. :)

File: AmplifyPlugins/DataStore/AWSDataStoreCategoryPlugin/Storage/StorageEngine+DeleteTransaction.swift (hunk 6 above)
[the remark this answers is not in the document]
Comment (palpatim) [reply AmplifyPlugins/DataStore/AWSDataStoreCategoryPlugin/Storage/StorageEngine+DeleteTransaction.swift:81]: Some kind of documentation of intent would be nice: is the 950 because we have some upper system bound? We feel it's a reasonable performance throttle? We've tested and this is the upper limit before the code falls over?

File: AmplifyPlugins/DataStore/AWSDataStoreCategoryPlugin/Storage/StorageEngine+DeleteTransaction.swift (hunk 8 above)
[the remark this answers is not in the document]
Comment (palpatim) [reply AmplifyPlugins/DataStore/AWSDataStoreCategoryPlugin/Storage/StorageEngine+DeleteTransaction.swift]: Yep, that's it.

File: AmplifyPlugins/DataStore/AWSDataStoreCategoryPlugin/Storage/StorageEngine.swift [hunk 9]
```diff
@@ -382,29 +389,39 @@ final class StorageEngine: StorageEngineBehavior {
                 return
             }
         }
-
+        var mutationEventsSubmitCompleted = 0
+        let serialQueue = DispatchQueue(label: "com.amazoncom.StorageEngine.syncDeletions.MType.concurrency")
```
Comment (palpatim) [inline AmplifyPlugins/DataStore/AWSDataStoreCategoryPlugin/Storage/StorageEngine.swift:393]: Serial queue makes sense to synchronize that whole block, but I don't know that we want to create it in this method. Why not make it a private property of the storage engine so we're not thrashing the system with dispatch queue creations? (We haven't yet done our [audit/refactor of queues](https://github.com/aws-amplify/amplify-ios/issues/139) to set up target queues per subsystem, so you don't need to set targets just yet.)

File: AmplifyPlugins/DataStore/AWSDataStoreCategoryPlugin/Storage/StorageEngine+DeleteTransaction.swift (hunk 6 above)
[the remark this answers is not in the document]
Comment (wooj2) [reply AmplifyPlugins/DataStore/AWSDataStoreCategoryPlugin/Storage/StorageEngine+DeleteTransaction.swift:81]: Ah, got it, I will add a comment.

**Review by palpatim — APPROVED** [review]
